### PR TITLE
Make calendars work with static data slicing and update PluralRules constructors

### LINKED ISSUE #1461: Make calendars work better with static data slicing
When I run the new `icu4x-key-extract` tool on `work_log`, I get

```
datetime/lengths@1
datetime/skeletons@1
datetime/symbols@1
plurals/ordinal@1
```

This is correct (the plurals are presumably pulled in for counting weeks).  However, there is no indication about which calendar system to use, even though a calendar system is explicitly specified:

```rust
let dtf = DateTimeFormat::<Gregorian>::try_new(locale, &provider, &options.into())
```

This is "by-design", of course: I expect the general case to be an `AnyCalendar`.  However, for cases where one particular calendar is being used, ideally we can get that info from the static data slicing tool.

Here's one idea.  Once data deduplication is implemented (#838), we can have multiple resource paths pointing at the same data, with no impact to data size (except for the index array).  When a calendar-specific DTF is created, it loads the calendar-specific key, and when an AnyCalendar DTF is created, it loads the all-calendar key.

This is not too different from the approach we're taking in Unicode Properties.  We have both pre-computed sets as well as all-inclusive CPTs, and it's expected that only one of those keys will be shipped.

Thoughts?  Any other ways to solve this problem?

- [x] @Manishearth 
- [x] @zbraniecki

### LINKED ISSUE #1571: Finalize PluralRules constructor structure
In https://github.com/unicode-org/icu4x/pull/1554 I'm adding new functions `PluralRules::try_new_ordinal` and `try_new_cardinal`.  We should discuss this at more length outside the context of https://github.com/unicode-org/icu4x/pull/1554.

The reason we need to split the constructor into an ordinal version and a cardinal version is for data slicing to work over FFI.  Since FFI needs separate functions, I think we should probably do the same in Rust.  I don't think `PluralRulesType` leads to significantly better ergonomics, so we should err on the side of consistency.

I think we should consider:

1. Get rid of `PluralRules::try_new` and `PluralRulesType`
2. Rename `PluralRules::try_new_cardinal` to `PluralRules::try_new` (assuming that cardinal is the much more common use case)

Needs approval from:

- [x] @zbraniecki

### LINKED ISSUE #2313: Rename Japanext to JapaneseExtended
See https://github.com/unicode-org/icu4x/pull/2311

### PR DESCRIPTION
**Pull request by sffc**
Also renames Japanext -> JapaneseExtended

Also updates PluralRules constructors

Fixes #2313
Fixes #1461 
Fixes #1571

### CODE REVIEW

**Review by Manishearth — DISMISSED** [review]

**Comment by sffc** [comment]
I had to datagen for databake after the JapaneseExtended change, so I decided to work off the same branch and finish everything else needed for #1461.

Looks like there are still some test failures. Will continue resolving those, but the rest of this should be ready for review. You can review commit-by-commit if you like in order to filter out the datagen-only commmits.

**Comment by sffc** [comment]
It works :-)

work_log.rs has

```rust
    let dtf = TypedDateTimeFormatter::<Gregorian>::try_new(
        &provider,
        &locale!("en").into(),
        &options.into(),
    )
```

and DCE leaves behind just `GregorianDateLengthsV1` and `GregorianDateSymbolsV1`.

I didn't touch the skeletons key because we need to re-do that one anyway.

**Comment by sffc** [comment]
I keep scope-creeping this PR instead of making new PRs because I don't want to needlessly resolve merge conflicts tomorrow

File: components/datetime/src/provider/calendar/mod.rs [hunk 1]
```diff
@@ -15,19 +15,23 @@ use icu_provider::{yoke, zerofrom};
 pub use skeletons::*;
 pub use symbols::*;
 
-#[icu_provider::data_struct(marker(
-    DateLengthsV1Marker,
-    "datetime/datelengths@1",
-    extension_key = "ca"
-))]
+#[icu_provider::data_struct(
+    marker(GregorianDateLengthsV1Marker, "datetime/gregory/datelengths@1"),
```
Comment (Manishearth) [inline components/datetime/src/provider/calendar/mod.rs:19]: thought: we can use this opportunity to call it datetime/gregorian, perhaps? Ideally we don't have to have the "gregory" name anywhere that is not BCP-47 constrained

File: components/datetime/src/provider/calendar/mod.rs [hunk 2]
```diff
@@ -15,19 +15,23 @@ use icu_provider::{yoke, zerofrom};
...
+    marker(BuddhistDateLengthsV1Marker, "datetime/buddhist/datelengths@1"),
+    marker(JapaneseDateLengthsV1Marker, "datetime/japanese/datelengths@1"),
+    marker(JapaneseExtendedDateLengthsV1Marker, "datetime/japanext/datelengths@1"),
+    marker(CopticDateLengthsV1Marker, "datetime/coptic/datelengths@1"),
+    marker(IndianDateLengthsV1Marker, "datetime/indian/datelengths@1"),
+    marker(EthiopicDateLengthsV1Marker, "datetime/ethiopic/datelengths@1")
```
Comment (Manishearth) [inline components/datetime/src/provider/calendar/mod.rs:25]: suggestion/thought: `datelengths/calendar` or `calendar/datelengths`? I feel like the former is nicer but I'm not sure.

The former gives a clear expectation that all calendars have the same formatting data key structure.

**Review by Manishearth — APPROVED** [review]
macro feels a bit gnarly but lgtm

File: components/datetime/src/provider/calendar/mod.rs (hunk 2 above)
Comment (sffc) [reply components/datetime/src/provider/calendar/mod.rs:25]: Could go either way. With calendar first, it means you get one directory per calendar, and the "foo@1" is the same as the data struct name FooV1, which is nice.

File: components/datetime/src/provider/calendar/mod.rs (hunk 1 above)
Comment (sffc) [reply components/datetime/src/provider/calendar/mod.rs:19]: I think I'd rather stick to the official identifiers. There is "gregorian" in CLDR JSON but that's not really standard. It opens questions about whether ethioaa, japanext, etc should be expanded, too

**Review by Manishearth — APPROVED** [review]

**Pull request merged**